### PR TITLE
Add model-parallel layout maps for 14 GQA model families + fix key/value sharding crashes (incl. Gemma v1)

### PR DESCRIPTION
**Pull request by pctablet505**
## Summary

Adds `get_layout_map` + `test_distribution` for model-parallel (`ModelParallel`) sharding to 14 GQA model
families: Gemma (v1 fix), Gemma3, Gemma4, Qwen3, QwenMoe, Qwen3Moe, Qwen3_5, Qwen3_5Moe, Mistral, Mixtral, Phi3,
GptOss, T5Gemma, T5Gemma2, and Llama (fix). Regex-based `LayoutMap` rules over a `(batch, model)` mesh, following
the existing `LlamaBackbone`/`GemmaBackbone` pattern.

## Bugs fixed

- **GQA key/value kernels sharded on an axis too small to divide the mesh.** `num_key_value_heads` is
  independent of and often much smaller than `num_query_heads` (down to 1 for MQA) — sharding it crashes with
  `IndivisibleError` whenever a mesh dim doesn't evenly divide it. Fix: query stays sharded; key/value's
  kv-heads axis is left replicated (hidden_dim axis still shards via `model_dim`). Found in two einsum
  conventions (Gemma-style heads-on-axis-0, Mistral-style heads-on-axis-1) across all 14 models, including
  **Gemma v1** — missed in an earlier pass, found by an independent review, whose test had also been edited to
  mask the crash rather than catch it (both fixed).
- **Missing `token_embedding/reverse_embeddings`** for models with untied embeddings (Mistral, Phi3, Qwen3,
  Llama, GptOss, Qwen3_5, Qwen3_5Moe, T5Gemma/T5Gemma2).

## Verification

- `test_distribution` passes for all 14 models at 2- and 4-device CPU mesh.
- Mesh-shape sweep across 2D (`2×4` .. `16×16`) and 3D (`2×2×2` .. `4×4×8`) topologies matching real TPU
  generations: **140/140** pass, verified per-weight.
- Cross-checked against **34 real production presets** (config-only, via Kaggle/HF) — zero regressions, since
  the fix only ever replicates (never adds a divisibility requirement).
- Independently verified on a real 8-device TPU mesh (Gemma4 26B MoE) — no compile-time errors or OOMs.
- `pre-commit run --all-files` clean.

## Known limitations (not addressed here)

- Query-head sharding still requires exact divisibility by the mesh width — pre-existing, inherent to tensor
  parallelism, affects several small real presets at 16-way sharding.
- Key/value are always replicated on the kv-heads axis, never conditionally sharded even when divisible (keeps
  the map static/mesh-agnostic; costs some unused parallelism headroom).
- Vision/audio/per-layer-input weights (Gemma3/4, Qwen3_5) and MoE expert-count axis (6 MoE models) are left
  replicated — separate scope.
- `QwenMoe`'s dense-FFN fallback layout is correct but untestable due to a pre-existing, unrelated `layer_index`
  bug (documented in-code, not fixed here).
- `QwenBackbone` (v1, non-MoE) likely has the same bug and wasn't audited; `Gemma3n`/`SmolLM3` use a
  structurally different (Dense-fused) attention kernel and aren't covered.
- JAX-only, matching the rest of keras-hub's distribution support.

### CODE REVIEW

File: keras_hub/src/models/gemma3/gemma3_backbone.py [hunk 1]
```diff
@@ -467,6 +467,76 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/gemma3/gemma3_backbone.py:493]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/gemma4/gemma4_backbone.py [hunk 2]
```diff
@@ -773,6 +773,100 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/gemma4/gemma4_backbone.py:801]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/gpt_oss/gpt_oss_backbone.py [hunk 3]
```diff
@@ -217,3 +217,100 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/gpt_oss/gpt_oss_backbone.py:243]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/mistral/mistral_backbone.py [hunk 4]
```diff
@@ -186,3 +186,83 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/mistral/mistral_backbone.py:212]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/mixtral/mixtral_backbone.py [hunk 5]
```diff
@@ -203,3 +203,87 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/mixtral/mixtral_backbone.py:229]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/phi3/phi3_backbone.py [hunk 6]
```diff
@@ -207,3 +207,83 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/phi3/phi3_backbone.py:233]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/qwen3/qwen3_backbone.py [hunk 7]
```diff
@@ -187,3 +187,83 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/qwen3/qwen3_backbone.py:213]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/qwen3_5/qwen3_5_backbone.py [hunk 8]
```diff
@@ -360,6 +360,94 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/qwen3_5/qwen3_5_backbone.py:388]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_backbone.py [hunk 9]
```diff
@@ -355,6 +355,121 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_backbone.py:383]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/t5gemma/t5gemma_backbone.py [hunk 10]
```diff
@@ -362,3 +362,105 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/t5gemma/t5gemma_backbone.py:389]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/t5gemma2/t5gemma2_backbone.py [hunk 11]
```diff
@@ -531,6 +531,113 @@ def get_config(self):
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras_hub/src/models/t5gemma2/t5gemma2_backbone.py:562]: Fixed — Args now specify type per the style guide, "partition" -> "partitioned", and `Return:` -> `Returns:`.

File: keras_hub/src/models/gemma3/gemma3_backbone.py [hunk 12]
```diff
@@ -467,6 +467,76 @@ def get_config(self):
...
+    @staticmethod
+    def get_layout_map(
+        device_mesh,
+        model_parallel_dim_name="model",
+        data_parallel_dim_name="batch",
+    ):
+        """Get a `keras.distribution.LayoutMap` for model parallel distribution.
+
+        The returned `LayoutMap` contains the sharding spec for the Gemma3
+        backbone text-decoder weights. Vision encoder weights are left
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/gemma3/gemma3_backbone.py:479]: >Vision encoder weights are left replicated 
- this has no test coverage

Comment (divyashreepathihalli) [reply keras_hub/src/models/gemma3/gemma3_backbone.py:479]: and why is there no sharding supported for vision encoder. replicating it on all teh device will increase the memory cost

File: keras_hub/src/models/gemma3/gemma3_backbone_test.py [hunk 13]
```diff
@@ -217,3 +218,64 @@ def test_all_presets(self):
...
+            self.skipTest("`ModelParallel` testing requires multiple devices.")
+        devices = devices[:2]
+        device_mesh = keras.distribution.DeviceMesh(
+            shape=(1, 2),
+            axis_names=("batch", "model"),
+            devices=devices,
+        )
+
+        layout_map = Gemma3Backbone.get_layout_map(device_mesh)
+        distribution = keras.distribution.ModelParallel(layout_map=layout_map)
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/gemma3/gemma3_backbone_test.py:236]: the test asserts sharding spec but not forward pass

**Review by divyashreepathihalli — COMMENTED** [review]
@pctablet505 for next time, please review go/small-cls

reviewing large changes takes a lot of time. please split your PRs to smallers ones next time

**Comment by divyashreepathihalli** [comment]
Please split this into smaller PRs. create a new PR for each model and make sure your layout map configurations work for each model by testing the training on a multi device setup. we can go from there. Thanks!

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

**Comment by pctablet505** [comment]
Splitting it into Multiple PRs
